go/vt/srvtopo/resolver.go

```diff
@@ -229,150 +160,31 @@ func (r *Resolver) GetAllKeyspaces(ctx context.Context) ([]string, error) {
-// and only if its corresponding key-space ids are in one of the key-ranges.
-func (r *Resolver) ResolveKeyRanges(ctx context.Context, keyspace string, tabletType topodatapb.TabletType, krs []*topodatapb.KeyRange) ([]*ResolvedShard, error) {
-	keyspace, _, allShards, err := r.GetKeyspaceShards(ctx, keyspace, tabletType)
-	if err != nil {
-		return nil, err
-	}
-	var res []*ResolvedShard
-	visited := make(map[string]bool)
-	for _, kr := range krs {
-		for _, shard := range allShards {
-			if !key.KeyRangesIntersect(kr, shard.KeyRange) {
-				// We don't need that shard.
-				continue
-			}
-			if visited[shard.Name] {
-				// We've already added that shard.
-				continue
-			}
-			// We need to add this shard.
-			visited[shard.Name] = true
-			target := &querypb.Target{
-				Keyspace:   keyspace,
-				Shard:      shard.Name,
-				TabletType: tabletType,
-				Cell:       r.localCell,
-			}
-			_, qs, err := r.stats.GetAggregateStats(target)
-			if err != nil {
-				return nil, resolverError(err, target)
-			}
-
-			// FIXME(alainjobart) we ignore the stats for now.
-			// Later we can fallback to another cell if needed.
-			// We would then need to read the SrvKeyspace there too.
-			target.Cell = ""
-			res = append(res, &ResolvedShard{
-				Target:       target,
-				QueryService: qs,
-			})
-		}
-	}
-	return res, nil
-}
-
-// ResolveEntityIds returns a list of shards and values to use in that shard.
-func (r *Resolver) ResolveEntityIds(ctx context.Context, keyspace string, entityIds []*vtgatepb.ExecuteEntityIdsRequest_EntityId, tabletType topodatapb.TabletType) ([]*ResolvedShard, [][]*querypb.Value, error) {
+// ResolveDestinations resolves values and their destinations into their
+// respective shards.
+//
+// If ids is nil, the returned [][]*querypb.Value is also nil.
+// Otherwise, len(ids) has to match len(destinations), and then the returned
+// [][]*querypb.Value is populated with all the values that go in each shard,
+// and len([]*ResolvedShard) matches len([][]*querypb.Value).
+//
+// Sample input / output:
+// - destinations: dst1, dst2, dst3
+// - ids:          id1,  id2,  id3
+// If dst1 is in shard1, and dst2 and dst3 are in shard2, the output will be:
+// - []*ResolvedShard:   shard1, shard2
+// - [][]*querypb.Value: [id1],  [id2, id3]
+func (r *Resolver) ResolveDestinations(ctx context.Context, keyspace string, tabletType topodatapb.TabletType, ids []*querypb.Value, destinations []key.Destination) ([]*ResolvedShard, [][]*querypb.Value, error) {
 	keyspace, _, allShards, err := r.GetKeyspaceShards(ctx, keyspace, tabletType)
 	if err != nil {
 		return nil, nil, err
 	}
+
 	var result []*ResolvedShard
 	var values [][]*querypb.Value
 	resolved := make(map[string]int)
-	for _, eid := range entityIds {
-		shard, err := GetShardForKeyspaceID(allShards, eid.KeyspaceId)
-		if err != nil {
-			return nil, nil, err
-		}
-		i, ok := resolved[shard]
+	addShard := func(shard string, i int) error {
```

Comment:
This is the first extensive use of local closures like this that I've seen in the Vitess code, certainly in a hot path.

I'm mildly curious whether there are any performance considerations to doing things this way as opposed to splitting each of these into bespoke functions and passing pointers to resolved, values, etc.

Reply:
I went over the top here. This code structure is a red flag that it needs refactoring, I was thinking about it too, in slack:

I think if I add a Resolve method in the Destination interface:
Resolve(allShards []*topodatapb.ShardReference, addShard func(shard string))
I can remove the big type check (and the closures!), and instead let it be just a call to that method.

Let me change the PR to that effect (although probably not any time soon, I need to go for the day soon, and I'm out tomorrow and Friday)